fix(gatsby-image): add required check for one of fluid or fixed

ehrencrona opened the pull request:
## Description

Prevents the rendering from crashing if Gatsby image is called without any parameters. The error message that was printed before this change was not helpful to users.

## Related Issues

Fixes #25370

polarathene commented on packages/gatsby-image/src/index.js:
```diff
@@ -74,9 +74,9 @@ const matchesMedia = ({ media }) =>
 const getImageSrcKey = ({ fluid, fixed }) => {
-  const data = fluid ? getCurrentSrcData(fluid) : getCurrentSrcData(fixed)
+  const currentData = fluid || fixed
 
-  return data.src
+  return currentData && getCurrentSrcData(currentData).src
```
Is this where the error is actually occurring, or within the function call logic itself?

I see `getCurrentSrcData()` use a conditional, with the 2nd condition being a method that runs checks, those would abort and not trigger further errors, except for the end of the function, where it's assumed an array has been provided to the function. The other method callers are in the fluid/fixed condition checks in render method, they should be fine.

The error here is from trying to access the `src` property? Your fix will return false otherwise, where `getImageSrcKey()` would be affected when creating a cache entry under `false` key. This can apparently be a valid key for an object, which may cause a problem if several `<Img />` utilize the same cache key?(I'm not sure if the cache object is shared among instances)

https://github.com/gatsbyjs/gatsby/blob/6102456c4a9f0a09aad5e3c2d61d6a38591ded2d/packages/gatsby-image/src/index.js#L120

If that would be an issue we'd want to address that.

---

I think this is more clear for where guard is appropriate:

```js
  const currentData = getCurrentSrcData(fluid || fixed)
  return currentData && currentData.src
```

Probably still want to avoid caching `false` as a cache key though if that's returned instead.

---

https://github.com/gatsbyjs/gatsby/blob/6102456c4a9f0a09aad5e3c2d61d6a38591ded2d/packages/gatsby-image/src/index.js#L103-L104

`hasArtDirectionSupport()` is only called by one method atm, the method that ends with the above under the assumption there is an array of objects provided.

https://github.com/gatsbyjs/gatsby/blob/6102456c4a9f0a09aad5e3c2d61d6a38591ded2d/packages/gatsby-image/src/index.js#L58-L59

You could add this to the start of `getCurrentSrcData()` to abort early with false bool.

---

Alternatively, is this something that the component should actually support? (No fluid or fixed object provided) Or should an error be thrown here(or earlier) instead? (albeit with a friendlier message)

ehrencrona replied:
@polarathene I agree a warning would be nicer! I hesitated to add one because I didn't want to bloat the client code. But I see now there's another warning in the image plugin that's gated by `process.env.NODE_ENV !== "production"`. I presume webpack removes that in production bundles. I'll do the same in this case then!

The error was `getCurrentSrcData(null)` crashing, but I didn't feel it reasonable that it should be able to handle being passed `null`. 

By the way, `null && whatever` evaluates to `null`, not `false`. You're right it's weird that `null` is later used as a cache key but I don't think it should cause any harm. 

Should i surround `imageCache[src] = true;` with an `if (src != null)` in `activateCacheForImage`? That way it's clear it's a non-cacheable value.

For better readability we could replace `null` with a constant `const invalidSrcData = null`, but that depends on your coding style.

polarathene replied:
> The error was getCurrentSrcData(null) crashing, but I didn't feel it reasonable that it should be able to handle being passed null. And you would need to check for whether src is present if it does.

The stackoverflow error suggested it tried to access `currentData.0` as a value to return as no array was provided to index. The method is meant to return an object, not just for the `src` property that this particular bug clearly has no access to.

> Should i surround `imageCache[src] = true;` with an `if (src != null)` in `activateCacheForImage`? That way it's clear it's a non-cacheable value.

Yes, that'd probably be best, but just `if (src) { ... }` should be fine, while unlikely to be an issue, it guards against all falsy values, whereas `src != null` only covers `null` and `undefined`, plus it's more terse so win-win :)

> but that depends on your coding style.

Personally I prefer the `&&` short-circuit approach you've used. It's already used quite heavily in the file, so any contributors hopefully know that a [falsy value](https://developer.mozilla.org/en-US/docs/Glossary/Falsy) can be returned. 

There's also some [double bang(NOT)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_NOT#Examples) operator usage, in one case at least it's to ensure a boolean value is provided for IE as it couldn't handle `undefined` properly if I remember right.

blainekasten replied:
> @polarathene I agree a warning would be nicer! I hesitated to add one because I didn't want to bloat the client code. But I see now there's another warning in the image plugin that's gated by process.env.NODE_ENV !== "production". I presume webpack removes that in production bundles. I'll do the same in this case then!
> 
> About throwing an error instead: I suppose it's conceivable that this happens at runtime (after deployment) if you pass a variable that has a dynamic, calculated value. At runtime you wouldn't want to crash. But that's kind of a code standard thing and I don't know how you do error handling elsewhere. And maybe it's far-fetched. If that's not a consideration, halting with an error is better.

Yeah, let's make sure to log a warning here. Let's not throw an error but just log one. So something like:

```js
if (process.env.NODE_ENV !== `production`) {
  if (!currentData) {
   console.error(`gatsby-image requires a valid src property. (maybe more actionable info here..)`)
  }
}
```

blainekasten replied:
I wonder if we could just wrap this in a `try/catch`

polarathene replied:
A `try/catch` would be good, perhaps with the following error message:

```
gatsby-image expected to receive a valid 'fixed' or 'fluid' prop, missing 'src' value.
```

If there isn't a valid reason for not supplying the image data prop(fixed / fluid), then the error could probably be fired off in the constructor. It's not going to do SSR any good either I imagine, the constructor will fire off on initial load too where adding some props will have no affect. 

If there's no intention to render the component initially, shouldn't they instead use conditional rendering? `{someCondition && <Img fluid={ ... } />}`.

LekoArts commented:
Coming from https://github.com/gatsbyjs/gatsby/issues/24855 - can you please add a test case where you pass in `null` to the fluid/fixed props?

polarathene commented:
> can you please add a test case where you pass in `null` to the fluid/fixed props?

That should have the same impact as `undefined`(prop not specified), as `null` will be a falsy type value too.

The error here is caused in the images constructor due to this call:

https://github.com/gatsbyjs/gatsby/blob/6102456c4a9f0a09aad5e3c2d61d6a38591ded2d/packages/gatsby-image/src/index.js#L333-L335

https://github.com/gatsbyjs/gatsby/blob/6102456c4a9f0a09aad5e3c2d61d6a38591ded2d/packages/gatsby-image/src/index.js#L110-L115

Not having a fluid/fixed object to trigger the render logic leaves just that afaik, or if using the critical prop, I think it's possible to trigger this via `handleImageLoaded()`:

https://github.com/gatsbyjs/gatsby/blob/6102456c4a9f0a09aad5e3c2d61d6a38591ded2d/packages/gatsby-image/src/index.js#L117-L122

If you're going to test with `null`, may also want to cover that case.

ehrencrona commented:
Wow, that triggered a lively debate! :D Thanks for all the feedback.

I've updated the PR with a warning. The warning only appears in development; I've checked that the code gets tree-shaken out in production builds.

I also applied the feedback from @polarathene about making sure the value is not cached.

Given #24855 it seems like this is something that can appear at runtime so we shouldn't bomb the page if it happens. The test ought to guarantee that.

Using `null` or `undefined` (=omitting the prop) in the test indeed ought to be equivalent. I switched to `null` since it's more explicit.

polarathene commented:
> (It would, by the way, be nice to pull those two `if` blocks out into their own functions to make the rendering logic more readable, but I didn't want to include that in this PR)

There's a fair amount of refactoring that needs to be done on `gatsby-image`. Some to better prep for v3, and then moving from a class component to hooks based for v3. No need to refactor those functions out of render, there's a number of PRs for this component waiting on being reviewed/merged, the less conflicts introduced the better :)

polarathene commented on packages/gatsby-image/src/index.js:
```diff
@@ -71,12 +71,12 @@ const matchesMedia = ({ media }) =>
  * Use `the first image in either `fixed` or `fluid`
  * @param {{fluid: {src: string, media?: string}[], fixed: {src: string, media?: string}[]}} args
- * @return {string}
+ * @return {string} or null if not cacheable
```
`[][0]` is `undefined` not `null`?

ehrencrona replied:
You're right! My bad. Updating.

wardpeet commented:
Hey @ehrencrona,

Thanks for creating this PR. Could we move this validation to propTypes?

```js
function requireFixedOrFluid(originalPropTypes) {
  return (props, propName, componentName) => {
    if (!props.fixed && !props.fluid) {
      throw new Error(`The prop \`fluid\` or \`fixed\` is marked as required in \`${componentName}\`, but their values are both \`undefined\`.`)
    }

    PropTypes.checkPropTypes({[propName]: originalPropTypes}, props, 'prop', componentName);
  }
}

// If you modify these propTypes, please don't forget to update following files as well:
// https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-image/index.d.ts
// https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-image/README.md#gatsby-image-props
// https://github.com/gatsbyjs/gatsby/blob/master/docs/docs/gatsby-image.md#gatsby-image-props
Image.propTypes = {
  resolutions: fixedObject,
  sizes: fluidObject,
  fixed: requireFixedOrFluid(PropTypes.oneOfType([fixedObject, PropTypes.arrayOf(fixedObject)])),
  fluid: requireFixedOrFluid(PropTypes.oneOfType([fluidObject, PropTypes.arrayOf(fluidObject)])),
  fadeIn: PropTypes.bool,
  durationFadeIn: PropTypes.number,
  title: PropTypes.string,
  alt: PropTypes.string,
  className: PropTypes.oneOfType([PropTypes.string, PropTypes.object]), // Support Glamor's css prop.
  critical: PropTypes.bool,
  crossOrigin: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  style: PropTypes.object,
  imgStyle: PropTypes.object,
  placeholderStyle: PropTypes.object,
  placeholderClassName: PropTypes.string,
  backgroundColor: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  onLoad: PropTypes.func,
  onError: PropTypes.func,
  onStartLoad: PropTypes.func,
  Tag: PropTypes.string,
  itemProp: PropTypes.string,
  loading: PropTypes.oneOf([`auto`, `lazy`, `eager`]),
  draggable: PropTypes.bool,
}
```

ehrencrona commented:
@wardpeet that would make sense of course; i can add that. but i'd say it should be in addition to the check in the PR, not a replacement for it, as the issue can also appear at runtime by a variable passed as `fluid` or `fixed` becoming null.

wardpeet commented:
Why wouldn't this be a replacement for it? I'm only talking about the validation part. The other changes you made are solid 💯.

ehrencrona commented:
@wardpeet Oh, I thought proptypes were only evaluated at compile time. Ok, no, if they're evaluated at runtime it's a strictly better solution of course. I'll update the code.

polarathene commented:
The propTypes validation is for development only and will work at run-time afaik, when built for production like the validation check code being wrapped in a conditional to exclude it, the propTypes aren't present either.

ehrencrona commented:
Ok, I've updated the code. Seems to work well.

wardpeet commented on packages/gatsby-image/src/index.js:
```diff
@@ -71,12 +71,12 @@ const matchesMedia = ({ media }) =>
  * Use `the first image in either `fixed` or `fluid`
  * @param {{fluid: {src: string, media?: string}[], fixed: {src: string, media?: string}[]}} args
- * @return {string}
+ * @return {string} or undefined if not cacheable
```
```suggestion
 * @return {string?} Returns image src or undefined it not given.
```

polarathene commented:
@wardpeet Please merge? :)

polarathene commented:
@wardpeet Are the approved changes going to be merged?

nicktouchette commented:
Should the required check extend to sizes as well?

Using something like
`<Img sizes={{ ...data.banner.childImageSharp.fluid, aspectRatio: 21 / 9 }} />`

Will give this error
`Warning: Failed prop type: The prop `fluid` or `fixed` is marked as required in `Image`, but their values are both `undefined`.`

polarathene commented:
`gatsby-image` isn't equivalent to `<img>` element. It does not have a supported [`sizes` prop](https://www.gatsbyjs.org/packages/gatsby-image/?=#gatsby-image-props), that's handled internally based on `fluid` or `fixed` prop being used.

There is the old `resolutions` and `sizes` props still technically supported, which last I knew were meant to state they're deprecated, doesn't seem like the component is logging that warning like it does for the `critical` prop. Seems it was discussed in [my old PR](https://github.com/gatsbyjs/gatsby/pull/14158/commits/844a496aefec6e46405c109563568c5b402f9004#r285505388), but didn't get a reply from @wardpeet where he resolved it unanswered instead.

nicktouchette commented:
@polarathene The reason I'm not using fluid is because I needed to enforce an aspect ratio of 1 on a batch of fluid images that I need to remain perfectly square.

The code piece I referenced comes directly from https://www.gatsbyjs.org/tutorial/gatsby-image-tutorial/ .

If deprecated, what would be the preferred method to enforcing an aspect ratio on fluid images?

polarathene commented:
@nicktouchette thanks for that link, I'll create a PR to update that, along with adding the deprecated notice for `sizes` and `resolutions`.

> If deprecated, what would be the preferred method to enforcing an aspect ratio on fluid images?

`sizes` works the same as `fluid`, so changing the prop alone should work.

nicktouchette commented:
@polarathene Nice!  I changed the prop back to fluid and everything's good.  Thanks for the insight.
